numpy/lib/function_base.py

```diff
@@ -4237,7 +4237,9 @@ def percentile(a,
     if a.dtype.kind == "c":
         raise TypeError("a must be an array of real numbers")
 
-    q = np.true_divide(q, 100)
+    # Use dtype of array if possible (e.g., if q is a python int or float)
+    # by making the divisor have the dtype of the data array.
+    q = np.true_divide(q, a.dtype.type(100) if a.dtype.kind == "f" else 100)
```

Comment:
Hmmm, feels a bit brittle, but I have to sleep on it probably (also doesn't generalize easily to user DTypes, but OK).

Maybe in this case it is actually easier to just track `was_pyscalar = type(q) in (int, float, complex)` and then apply `if was_pyscalar: pos = float(pos)` at the end of the calculation.

Beyond figuring this out, one poitn I am annoying about it is, that I think this is an important problem to get eyes on, because `percentile` can't be the only function in NumPy (and even more so downstream!) to run into this.

Reply:
The problem with `was_pyscalar` is that I think the expectation should be that the result keeps the array input `dtype`, but only if it is `float`.

Reply:
Ah, by `pos` there, I don't mean the end-result, but the interpolation point value that is used for the last bit of the calculation.  I.e. the intermediate result just before mixing it with the values from `a`.

Reply:
Ah, I see. Indeed, that could work too. I think I slightly prefer to trying to deal with things up front - I'm trying to think of these functions as `gufuncs` and following the same logic...